### PR TITLE
Feature/scc 441 create event

### PR DESCRIPTION
**Pull request by adrianmilne**
Will create an Rfx/Event at the same time as a Project now.

Also can create directly via a POST to `{{baseUrl}}/tenders/projects/1/events` (with a body of {}).

Sure there will be some code tidying to pick up on 😉

### CODE REVIEW

File: src/main/java/uk/gov/crowncommercial/dts/scale/cat/model/entity/ProcurementEvent.java [hunk 1]
```diff
@@ -45,4 +53,29 @@
...
+   * Builds an instance from basic details
+   *
+   * @param jaggaerEventId The rfx reference code
+   * @param principal
+   * @return a procurement event
+   */
+  public static ProcurementEvent of(Integer projectId, String eventName, String jaggaerEventId,
+      String ocdsAuthority, String ocidPrefix, String principal) {
+    var procurementEvent = new ProcurementEvent();
+    procurementEvent.setProjectId(projectId);
```
Comment (tommyb82) [inline src/main/java/uk/gov/crowncommercial/dts/scale/cat/model/entity/ProcurementEvent.java:66]: Would it be preferable to add the new `ProcurementEvent` to the `ProcurementProject` entity's collection of events, and have the persistence layer then handle setting the foreign key column value on a save to the ProcurementProject repo..? Dunno - just thinking I can't remember explicitly setting the FK value before..

File: src/main/java/uk/gov/crowncommercial/dts/scale/cat/config/OAuth2Config.java [hunk 2]
```diff
@@ -28,6 +28,7 @@ protected void configure(HttpSecurity http) throws Exception {
       authz
         // TODO: Update to correct roles when fixed in Auth server
         .antMatchers(HttpMethod.POST, "/tenders/projects/agreements").hasAuthority("[\"CAT_ADMINISTRATOR\",\"CAT_USER\"]")
+        .antMatchers(HttpMethod.POST, "/tenders/projects/*/events").hasAuthority("[\"CAT_ADMINISTRATOR\",\"CAT_USER\"]")
```
Comment (tommyb82) [inline src/main/java/uk/gov/crowncommercial/dts/scale/cat/config/OAuth2Config.java:31]: I don't know how granular we want / need to be with these, going forward.. could combine into a single `/tenders/*` pattern for now I guess 🤷

File: src/main/java/uk/gov/crowncommercial/dts/scale/cat/model/entity/ProcurementEvent.java [hunk 3]
```diff
@@ -45,4 +53,29 @@
...
+    procurementEvent.setJaggaerEventId(jaggaerEventId);
+    procurementEvent.setCreatedBy(principal); // Or Jaggaer user ID?
+    procurementEvent.setCreatedAt(Instant.now());
+    procurementEvent.setUpdatedBy(principal); // Or Jaggaer user ID?
+    procurementEvent.setUpdatedAt(Instant.now());
+    return procurementEvent;
+  }
+
+  public String getEventID() {
+    return ocidprefix + "-" + ocdsAuthorityName + "-" + id;
```
Comment (tommyb82) [inline src/main/java/uk/gov/crowncommercial/dts/scale/cat/model/entity/ProcurementEvent.java:79]: Swap round - should be `authority-prefix-id` (I think - https://standard.open-contracting.org/latest/en/schema/identifiers/ 🤓

File: src/test/java/uk/gov/crowncommercial/dts/scale/cat/service/ProcurementEventServiceTest.java [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+    assertEquals(OCID_PREFIX, captor.getValue().getOcidprefix());
+    assertEquals(OCDS_AUTH_NAME, captor.getValue().getOcdsAuthorityName());
+    assertEquals(RFX_REF_CODE, captor.getValue().getJaggaerEventId());
+    assertEquals(PROC_PROJECT_ID, captor.getValue().getProjectId());
+    assertEquals(PRINCIPAL, captor.getValue().getCreatedBy());
+    assertNotNull(captor.getValue().getCreatedAt());
+    assertNotNull(captor.getValue().getUpdatedAt());
+
+    // Verify that event ID is generated correctly by entity
+    assertEquals(OCID_PREFIX + "-" + OCDS_AUTH_NAME + "-1", procurementEvent.getEventID());
```
Comment (tommyb82) [inline src/test/java/uk/gov/crowncommercial/dts/scale/cat/service/ProcurementEventServiceTest.java:122]: At least the test is consistent... 😜  (swap OCDS args)

File: src/test/java/uk/gov/crowncommercial/dts/scale/cat/service/ProcurementEventServiceTest.java [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+
+    when(procurementEventRepo.save(any(ProcurementEvent.class))).then(mock -> {
+      procurementEvent.setId(PROC_PROJECT_ID);
+      procurementEvent.setOcdsAuthorityName(OCDS_AUTH_NAME);
+      procurementEvent.setOcidprefix(OCID_PREFIX);
+      return procurementEvent;
+    });
+
+    // Invoke
+    ArgumentCaptor<ProcurementEvent> captor = ArgumentCaptor.forClass(ProcurementEvent.class);
```
Comment (tommyb82) [inline src/test/java/uk/gov/crowncommercial/dts/scale/cat/service/ProcurementEventServiceTest.java:107]: Noice 👍

**Review by tommyb82 — CHANGES_REQUESTED** [review]
Fantastic - just the one concrete change to the Event OCID - other than that all optional / comments 👍

File: src/main/java/uk/gov/crowncommercial/dts/scale/cat/config/OAuth2Config.java (hunk 2 above)
Comment (adrianmilne) [reply src/main/java/uk/gov/crowncommercial/dts/scale/cat/config/OAuth2Config.java:31]: Yep - makes sense - I'll update that 👍

File: src/main/java/uk/gov/crowncommercial/dts/scale/cat/model/entity/ProcurementEvent.java (hunk 3 above)
Comment (adrianmilne) [reply src/main/java/uk/gov/crowncommercial/dts/scale/cat/model/entity/ProcurementEvent.java:79]: Nice one - swapping now 👍

File: src/main/java/uk/gov/crowncommercial/dts/scale/cat/model/entity/ProcurementEvent.java (hunk 1 above)
Comment (adrianmilne) [reply src/main/java/uk/gov/crowncommercial/dts/scale/cat/model/entity/ProcurementEvent.java:66]: What about if it was a proper bidirectional relationship? (which I think is what my original intention was - this probably was a first cut which just got left there - agree setting the foreignKey value explicitly isn't great - missed that)

Comment (tommyb82) [reply src/main/java/uk/gov/crowncommercial/dts/scale/cat/model/entity/ProcurementEvent.java:66]: Yep absolutely - or that, probably more use long term 👍

**Review by tommyb82 — APPROVED** [review]
Magnifico! 👍

**Pull request merged**